Kernel/Arch/riscv64/InterruptManagement.cpp

```diff
@@ -34,7 +36,64 @@ void InterruptManagement::initialize()
 
 void InterruptManagement::find_controllers()
 {
-    // TODO: Once device tree support is in place, find interrupt controllers using that.
+    auto const& device_tree = DeviceTree::get();
+    auto maybe_soc = device_tree.get_child("soc"sv);
+    if (!maybe_soc.has_value()) {
+        dmesgln("Interrupts: No `soc` node found in the device tree, Interrupts initialization will be skipped");
+        return;
+    }
+    auto const& soc = maybe_soc.value();
+    auto soc_address_cells = soc.get_property("#address-cells"sv).value().as<u32>();
+    auto soc_size_cells = soc.get_property("#size-cells"sv).value().as<u32>();
+    auto interrupt_controllers_seen = 0;
+    for (auto const& [node_name, node] : soc.children()) {
+        if (!node.has_property("interrupt-controller"sv))
+            continue;
+
+        interrupt_controllers_seen++;
+
+        auto maybe_compatible = node.get_property("compatible"sv);
+        if (!maybe_compatible.has_value()) {
+            dmesgln("Interrupts: Devicetree node for {} does not have a 'compatible' string, rejecting", node_name);
+            continue;
+        }
+
+        // Reject non sifive-compatible interrupt controllers
+        auto sifive_plic = false;
+        maybe_compatible->for_each_string([&sifive_plic](StringView compatible_string) -> IterationDecision {
+            if (compatible_string == "sifive,plic-1.0.0"sv) {
+                sifive_plic = true;
+                return IterationDecision::Break;
+            }
+            return IterationDecision::Continue;
+        });
+        if (!sifive_plic)
+            continue;
+
+        auto maybe_reg = node.get_property("reg"sv);
+        if (!maybe_reg.has_value()) {
+            dmesgln("Interrupts: Devicetree node for {} does not have a physical address assigned to it, rejecting", node_name);
+            continue;
+        }
+        auto reg = maybe_reg.value();
+        auto stream = reg.as_stream();
+        FlatPtr paddr;
+        if (soc_address_cells == 1)
+            paddr = MUST(stream.read_value<BigEndian<u32>>());
+        else
+            paddr = MUST(stream.read_value<BigEndian<u64>>());
+        size_t size;
+        if (soc_size_cells == 1)
+            size = MUST(stream.read_value<BigEndian<u32>>());
+        else
+            size = MUST(stream.read_value<BigEndian<u64>>());
+        auto max_interrupt_id = node.get_property("riscv,ndev"sv).value().as<u32>();
+        m_interrupt_controllers.append(adopt_lock_ref(*new (nothrow) PLIC(PhysicalAddress(paddr), size, max_interrupt_id + 1)));
```

Comment:
Are you sure that the `riscv,ndev` property doesn't include interrupt 0?
Because now that you changed the for loop to use `i <= (m_interrupt_count >> 5)` qemu is throwing these warnings: `sifive_plic_write: Invalid enable write 0x208c`.

Reply:
Ah no, the + 1 seems right.
But I think you need to do do `i <= ((m_interrupt_count - 1) >> 5)`?

Reply:
Yes, you're right, since it's a count.